File: Data Structures & Algorithms/buy-and-sell-crypto/submission-4.java
class Solution {
    public int maxProfit(int[] prices) {
        // 2 pointers
        // if (prices.length < 2) return 0;

        // int buylow = prices[0];
        // int maxprofit = 0;
        // for (int price: prices) {
        //     if (price > buylow) {
        //         int profit = price - buylow;
        //         maxprofit = Math.max(maxprofit, profit);
        //     } else {
        //         buylow = price;
        //     }
        // }
        // return maxprofit;

        // dynamic programming
        
        int maxprofit = 0;
        int buylow = prices[0];
        for (int price: prices) {
            maxprofit = Math.max(maxprofit, price - buylow);
            buylow = Math.min(buylow, price);
        }
        return maxprofit;
        
    }
}

File: Data Structures & Algorithms/max-water-container/submission-4.cpp
class Solution {
public:
    int maxArea(vector<int>& heights) {
        int l = 0, r = heights.size() -1;
        int maxi = 0;
        while (l < r) {
            int area = min(heights[l], heights[r]) * (r-l);
            maxi = max(maxi, area);
            if (heights[l]>heights[r]) {
                r--;
            } else {
                l++;
            }
        }
        return maxi;
    }
};
